### PR TITLE
Implemented closed-form solution in `detect_clearsky`

### LINKED ISSUE #2216: Simple closed-form solution for alpha in detect_clearsky - clarified
### Discussed in https://github.com/pvlib/pvlib-python/discussions/2171

<div type='discussions-op-text'>

<sup>Originally posted by **agodbehere** August 20, 2024</sup>
The algorithm in clearsky.detect_clearsky optimizes for a scaling parameter alpha by finding a value that minimizes the root mean-squared error (RMSE) between measurements and a scaled clear-sky model. Currently, the implementation computes the RMSE and uses scipy.optimize to iteratively minimize the function. However, a simple closed-form solution exists, bypassing the need to call scipy.optimize.

The objective function is $f(\alpha) = \sqrt{\frac{\sum_i (G_i - \alpha C_i)^2}{n}}$, where $G$ represents measurements of GHI and $C$ represents the clear-sky model.

Since $f(\alpha)$ is nonnegative, $\arg\min_\alpha f(\alpha) = \arg\min_\alpha f^2(\alpha)$. The value $n$ is a constant and similarly does not affect the solution. 

Thus, we end up with a modified objective function $\hat{f}(\alpha) = \sum_i (G_i-\alpha C_i)^2$. Solving for $\frac{d\hat{f}}{d\alpha} = 0$ yields the optimal value: $\hat{\alpha} = \frac{\sum_i G_iC_i}{\sum_i C_i^2}$. The result is nonnegative, so the constraint $\alpha > 0$ is satisfied implicitly. 
</div>

### PR DESCRIPTION
**Pull request by agodbehere**
- [x] Closes #2216 
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

**Review by echedey-ls — COMMENTED** [review]
Nice PR @agodbehere 💯  !!

Leaving early review comments:
- You can safely address flake8 warnings in `test_clearsky.py`
- Docs will need to be updated to reflect what new algorithm is being used, which I identify are:
  - The docstring of the modified function (`pvilb.clearsky.detect_clearsky`)
  - The user guide section at [`clearsky.rst`](https://github.com/pvlib/pvlib-python/blob/5d23f145efbe5b4ab44ceda5d29e16eecba288ff/docs/sphinx/source/user_guide/clearsky.rst?plain=1#L571)
  - As you update these sections, you will be able to see how they will look when published at the following link (you can access it at the bottom of this PR in "Checks", through the "ReadTheDocs" action): https://pvlib-python--2217.org.readthedocs.build/en/2217/
- I haven't checked the code much more deeply, but there may be parts relating to the old algorithm at other places too

Feel free to ask about whatever you need help with, we all know it's not that easy to contribute to open-source 🚀

File: docs/sphinx/source/whatsnew/v0.11.1.rst [hunk 1]
```diff
@@ -30,6 +30,10 @@ Enhancements
 * Added function for calculating wind speed at different heights,
   :py:func:`pvlib.atmosphere.windspeed_powerlaw`.
   (:issue:`2118`, :pull:`2124`)
+* Implemented closed-form solution for alpha in detect_clearsky, obviating
+  the call to scipy.optimize that was prone to runtime errors and minimizing
+  computation, :py:func:`pvlib.clearsky.detect_clearsky`. Resolves :issue:`2712`.
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply docs/sphinx/source/whatsnew/v0.11.1.rst:35]: Should be `:issue:`2171`, :issue:`2216`

File: pvlib/clearsky.py [hunk 2]
```diff
@@ -874,25 +873,11 @@ def detect_clearsky(measured, clearsky, times=None, infer_limits=False,
...
-                           update your scipy version (try version 1.8.0 \
-                           or beyond)."
-            raise RuntimeError(message)
-
-        else:
-            alpha = optimize_result.x
-
+        # Compute arg min of MSE between model and observations
+        C = (clear_clear**2).sum()
+        if not (pd.isna(C) or C == 0):  # safety check
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/clearsky.py:878]: Without the check the next line divides by `C`. The exit criteria need `alpha` to be a float. I think the check and `if...break` are fine.

File: pvlib/clearsky.py [hunk 3]
```diff
@@ -874,25 +873,11 @@ def detect_clearsky(measured, clearsky, times=None, infer_limits=False,
...
+            # only update alpha if C is strictly positive
+            alpha = (clear_meas * clear_clear).sum() / C
         if round(alpha*10000) == round(previous_alpha*10000):
             break
     else:
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/clearsky.py:883]: Definitely remove the `else` block.

File: docs/sphinx/source/whatsnew/v0.11.1.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (echedey-ls) [reply docs/sphinx/source/whatsnew/v0.11.1.rst:35]: Updated, thanks ❣️

File: pvlib/clearsky.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (agodbehere) [reply pvlib/clearsky.py:883]: I would argue that the else block needs to stay - 
The algorithm is a nested optimization. What I have changed is just the inner optimization problem. 

In the outer loop, a subset of the data is selected as "clear sky" given the current value of alpha. Given this subset of data selected as "clear sky", the updated optimization algorithm is run in order to find an updated alpha. This iteration continues until alpha converges. 

What do you think?

[the remark this answers is not in the document]
Comment (agodbehere) [reply pvlib/clearsky.py:883]: Similarly, I think the current docstring still applies:
"An iterative algorithm identifies clear periods, uses the identified periods to estimate bias in the clearsky data, scales the clearsky data and repeats." 
The "estimate bias in the clearsky data" is what I have changed from an iterative optimization to a closed-form solution.

[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/clearsky.py:883]: Ah, yes. I agree now with @agodbehere keep the `else` block. When paired with the `for` statement I mistakenly though the `raise` would always execute when the loop ends. But not when the loop ends with the `break` statement; in that case (convergence) it is skipped.

File: docs/sphinx/source/whatsnew/v0.11.1.rst (hunk 1 above)
Comment (RDaxini) [inline docs/sphinx/source/whatsnew/v0.11.1.rst:35]: Link the function in the first line and then no need to repeat it here at the end as (I think?) we only do that if it is a new function being added
```suggestion
* Implemented closed-form solution for alpha in
  :py:func:`pvlib.clearsky.detect_clearsky`, obviating
  the call to scipy.optimize that was prone to runtime errors and minimizing
  computation. (:issue:`2171`, :issue:`2216`, :pull:`2217`).
```

**Review by RDaxini — COMMENTED** [review]
Nice job 👍🏾
Minor suggestion for the whatsnew file (someone will correct me if I'm mistaken)

File: pvlib/clearsky.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (echedey-ls) [reply pvlib/clearsky.py:883]: Consider this conversation resolved then; thanks for the explanations guys!

**Review by cwhanse — APPROVED** [review]
I approve providing the comments on the whatsnew file are resolved.

**Review by RDaxini — APPROVED** [review]

**Pull request merged**